File: src/training/fitness.js
// Fitness formula and standalone bot evaluation against a pool.
import { runGame } from '../simulation/runner.js';

/**
 * Compute fitness from raw game stats.
 * Draws are penalised heavily — they count double in the denominator.
 * This discourages draw-farming strategies.
 *
 *   fitness = wins / (wins + losses + draws × 2)
 *
 * Returns 0 if no games have been played.
 *
 * @param {{ wins: number, losses: number, draws: number }} stats
 * @returns {number} fitness in [0, 1]
 */
export function calcFitness({ wins, losses, draws }) {
  const denominator = wins + losses + draws * 2;
  return denominator === 0 ? 0 : wins / denominator;
}

/**
 * Evaluate a single bot against every bot in `pool`.
 * Useful for one-off evaluation outside the tournament (e.g. testing Phase C bots).
 *
 * @param {{ id: string, fn: Function }} bot
 * @param {{ id: string, fn: Function }[]} pool
 * @param {number} gamesPerOpponent  - alternates first player
 * @returns {{ wins, losses, draws, fitness }}
 */
export function evaluateFitness(bot, pool, gamesPerOpponent = 10) {
  let wins = 0, losses = 0, draws = 0;

  for (const opponent of pool) {
    for (let g = 0; g < gamesPerOpponent; g++) {
      const { winner } = runGame(bot.fn, opponent.fn, { botAGoesFirst: g % 2 === 0 });
      if (winner === 'A')      wins++;
      else if (winner === 'B') losses++;
      else                     draws++;
    }
  }

  return { wins, losses, draws, fitness: calcFitness({ wins, losses, draws }) };
}

File: src/training/fitness.py
# Fitness formula and standalone bot evaluation against a pool.
from ..simulation.runner import run_game


def calc_fitness(wins, losses, draws):
    """
    Compute fitness from raw game stats.
    Draws are penalised heavily — they count double in the denominator.
    This discourages draw-farming strategies.

        fitness = wins / (wins + losses + draws * 2)

    Returns 0 if no games have been played.
    """
    denominator = wins + losses + draws * 2
    return 0 if denominator == 0 else wins / denominator


def evaluate_fitness(bot, pool, games_per_opponent=10):
    """
    Evaluate a single bot against every bot in `pool`.
    Useful for one-off evaluation outside the tournament.

    bot:  dict with 'id' and 'fn'
    pool: list of dicts with 'id' and 'fn'
    Returns: dict { wins, losses, draws, fitness }
    """
    wins = losses = draws = 0

    for opponent in pool:
        for g in range(games_per_opponent):
            result = run_game(bot['fn'], opponent['fn'], bot_a_goes_first=(g % 2 == 0))
            if result['winner'] == 'A':
                wins += 1
            elif result['winner'] == 'B':
                losses += 1
            else:
                draws += 1

    return {'wins': wins, 'losses': losses, 'draws': draws,
            'fitness': calc_fitness(wins, losses, draws)}
